module/ice-9/generic-hash-tables.scm

```diff
@@ -0,0 +1,854 @@
+                   ;; These procedures don't need to be wrapped.
+                   ((or (eq? string-hash hash-function)
+                        (eq? string-ci-hash hash-function))
+                    (values hash-function hash-function))
+                   ((or (eq? hashq hash-function)
+                        (eq? hash-by-identity hash-function))
+                    (values hashq hash-function))
+                   ((or (eq? hashv hash-function)
+                        (eq? hash-by-value hash-function))
+                    (values hashv hash-function))
+                   ;; Otherwise, wrap the given function.
+                   (else (values (wrap-hash-function hash-function) hash-function))))
+   ((eq? equal? equiv-function) (values (@ (guile) hash) hash))
+   ((eq? eq? equiv-function) (values hashq hash-by-identity))
+   ((eq? eqv? equiv-function) (values hashv hash-by-value))
+   ((eq? string=? equiv-function) (values string-hash string-hash))
+   ((eq? string-ci=? equiv-function) (values string-ci-hash string-ci-hash))
+   (else (error "A suitable hash function could not be determined" equiv-function))))
+
+(define (guile-ht-ctor weakness)
+  "Answer the Guile HT constructor for the given WEAKNESS, where
+WEAKNESS is as specified by SRFI-126."
+  (case weakness
+    ((#f) (@ (guile) make-hash-table))
+    ((weak-key) make-weak-key-hash-table)
+    ((weak-value) make-weak-value-hash-table)
+    ((weak-key-and-value) make-doubly-weak-hash-table)
+    ((ephemeral-key ephemeral-value ephemeral-key-and-value)
+     (error "Unsupported hash table weakness" weakness))
+    (else (error "Invalid hash table weakness" weakness))))
+
+(define (equivalence-proc->associator equiv-function)
+  "Answer an `assoc'-like procedure that compares the argument key to
+alist keys with EQUIV-FUNCTION."
+  (cond ((or (eq? equal? equiv-function)
+	     (eq? string=? equiv-function)) (@ (guile) assoc))
+	((eq? eq? equiv-function) assq)
+	((eq? eqv? equiv-function) assv)
+	(else (lambda (item alist)
+		(assoc item alist equiv-function)))))
+
+(define-record-type generic-hash-table
+  (make-generic-hash-table real-table hash-function associator weakness
+                           mutable? equivalence-function original-hash-function)
+  hash-table?
+  ;; These three are the most accessed fields.
+  (real-table ht-real-table ht-real-table!)
+  (hash-function ht-hash-function)
+  (associator ht-associator)
+  ;; Weak hash tables don't use handles.
+  (weakness ht-weakness)
+  ;; Supports immutability.
+  (mutable? hash-table-mutable?)
+  ;; These are mostly needed for reflective queries
+  (equivalence-function hash-table-equivalence-function)
+  (original-hash-function hash-table-hash-function))
+
+;; Show some informations.
+(define (print-hash-table ht port)
+  (let ((equiv-name (procedure-name (hash-table-equivalence-function ht))))
+    (format port "#<generic-hash-table ~@[~a ~]~@[~a ~]size: ~a mutable? ~a>"
+            equiv-name (ht-weakness ht)
```

Comment:
here it's not clear to me what `ht-weakness` returns, also you prefix with `size:` and `mutable?:` but not but there is no prefix for `weakness`

Reply:
Yes, I agree `hash-table-weakness` should have a clear documentation, I can change that tomorrow. About printing, `weakness` can be `#f` (in which case it's not printed) or it is one of `weak-key`, `weak-value` or `weak-key-and-value`, so I think prefixing it with `weakness:` seems redundant to me IMHO.